crates/typst/src/foundations/calc.rs

```diff
@@ -908,6 +912,91 @@ pub fn quo(
     Ok(floor(dividend.apply2(divisor.v, Div::div, Div::div)))
 }
 
+/// Calculates the euclidean norm of a sequence of values.
+///
+/// ```example
+/// #calc.norm2(1, 2, -3, 0.5)
+/// #calc.norm2(1in, 2cm)
+/// #calc.norm2(3em, 4em)
+/// ```
+#[func(title = "Euclidean Norm")]
+pub fn norm2(
+    /// The sequence of values from which to calculate the euclidean norm.
+    #[variadic]
+    values: Vec<Spanned<Value>>,
+) -> SourceResult<Value> {
+    pnorm(Num::Int(2), values)
+}
+
+/// Calculates the p-norm of a sequence of values.
+///
+/// ```example
+/// #calc.pnorm(2, 1, 2, -3, 0.5)
+/// #calc.pnorm(3, 1in, 2cm)
+/// #calc.pnorm(1.5, 3em, 4em)
+/// ```
+#[func(title = "P-Norm")]
+pub fn pnorm(
+    /// The p value to calculate the p-norm of.
+    p: Num,
+    /// The sequence of values from which to calculate the p-norm.
+    /// Returns `0.0` if empty.
+    #[variadic]
+    values: Vec<Spanned<Value>>,
+) -> SourceResult<Value> {
+    let mut sum = 0.0;
+    if let Some(Spanned { v, span }) = values.first() {
+        match v {
+            Value::Int(_) | Value::Float(_) => {
+                for Spanned { v, span } in values {
+                    match v {
+                        Value::Int(n) => sum += (n as f64).powf(p.float()),
+                        Value::Float(n) => sum += n.powf(p.float()),
+                        _ => bail!(span, "expected a number"),
+                    }
+                }
+                Ok(Value::Float(sum.powf(1.0 / p.float())))
+            }
+            Value::Length(Length { em, .. }) if em.is_zero() => {
+                for Spanned { v, span } in values {
+                    match v {
+                        Value::Length(Length { abs, em }) if em.is_zero() => {
+                            sum += abs.to_raw().powf(p.float())
+                        }
+                        _ => {
+                            bail!(span, "expected an absolute length"; hint: "use `to-absolute()` to convert to an absolute length")
```

Comment:
In the current code base, it is recommended to break such a long line into multiple lines. For example:

```suggestion
                            bail!(
                                span, "expected an absolute length";
                                hint: "use `to-absolute()` to convert to an absolute length"
                            )
```